### ISSUE TITLE
Fix issue where alpha APIs that have `k8s:prerelease-lifecycle-gen:introduced` have an auto generated `APILifecycleRemoved` (should only be for beta/GA APIs)

### ISSUE THREAD

**Issue by aaron-prindle** [kubernetes/kubernetes#127249]
### What happened?

**EDIT: currently all the info in the issue here uses APIs generally but it should state alpha APIs as the current `APILifecycleRemoved` policy is being added to alpha APIs when it is unclear it should be and that is the root of the issue**

Currently when using the `// +k8s:prerelease-lifecycle-gen:introduced=<version>` tag on API types.go files, the associated generated code (`zz_generated.prerelease-lifecycle.go`) creates an APILifecycleRemoved method for that API automatically set for introduced + 6 minor versions.  Specifically, there are a number of APIs that have // +k8s:prerelease-lifecycle-gen:introduced=1.26 and this means that these APIs have APILifecycleRemoved methods targeting v1.32.  This means when attempting to bump `DefaultKubeBinaryVersion` from v1.31 -> v1.32 there are integration test failures where the test expects APIs to exist that were removed via the above flow (more details in other sections).  Below is an example of such an entry from `master` @ HEAD:
File: ./staging/src/k8s.io/api/authentication/v1alpha1/zz_generated.prerelease-lifecycle.go
```
// APILifecycleRemoved is an autogenerated function, returning the release in which the API is no longer served as int versions of major and minor for comparison.
// It is controlled by "k8s:prerelease-lifecycle-gen:removed" tags in types.go or  "k8s:prerelease-lifecycle-gen:deprecated" plus three minor.
func (in *SelfSubjectReview) APILifecycleRemoved() (major, minor int) {
	return 1, 32
```

For the integration tests, I have validated it is what [folks mentioned in the PR comments](https://github.com/kubernetes/kubernetes/pull/126977#issuecomment-2332596429) - the usage of `// +k8s:prerelease-lifecycle-gen:introduced=1.26"`

It seems for all APIs where there is used there is a APILifecycleRemoved method added in all of the related generated code with 1.32 as the version to remove that API.   This makes it so that the associated tests fail.  If you manually change these values to 1.33, the tests all pass.  
```
diff --git a/staging/src/k8s.io/api/authentication/v1alpha1/zz_generated.prerelease-lifecycle.go b/staging/src/k8s.io/api/authentication/v1alpha1/zz_generated.prerelease-lifecycle.go
index 62a70a781d1..af598cb4db1 100644
--- 
 // APILifecycleRemoved is an autogenerated function, returning the release in which the API is no longer served as int versions of major and minor for comparison.
 // It is controlled by "k8s:prerelease-lifecycle-gen:removed" tags in types.go or  "k8s:prerelease-lifecycle-gen:deprecated" plus three minor.
 func (in *SelfSubjectReview) APILifecycleRemoved() (major, minor int) {
-    return 1, 32
+    return 1, 33
 }
```

For further evidence this the root cause of the failing integration tests at https://github.com/kubernetes/kubernetes/pull/126977, below is the list of failing integration tests:
- TestGetsSelfAttributes
- TestCTBAttestPlugin
- TestCTBSignerNameFieldSelector
- TestCTBSignerNameChangeForbidden
- TestEncryptAll
- TestEtcdStoragePath
- TestAPIServerMetrics

and here are k8s types that APILifeCycleRemoved and APILifeCycleDeprecated set for 1.32:

APILifecycleDeprecated
- APIGroupDiscovery (from apidiscovery/v2beta1)
- APIGroupDiscoveryList (from apidiscovery/v2beta1)
- VolumeAttributesClass (from storage/v1alpha1)
- VolumeAttributesClassList (from storage/v1alpha1)

APILifecycleRemoved
- SelfSubjectReview (from authentication/v1alpha1)
- FlowSchema (from flowcontrol/v1beta3)
- FlowSchemaList (from flowcontrol/v1beta3)
- PriorityLevelConfiguration (from flowcontrol/v1beta3)
- PriorityLevelConfigurationList (from flowcontrol/v1beta3)
- ClusterTrustBundle (from certificates/v1alpha1)
- ClusterTrustBundleList (from certificates/v1alpha1)

See the gist here for the full code snippets showing the above methods:
https://gist.github.com/aaron-prindle/3e8a5c3cef3b8ff10763be0d4858254d

You can see that the k8s types above align 1:1 with the integration test failues stating that k8s objects don't exist.

In speaking offline w/ @jpbetz it seems a viable short-term solution to unblock PR https://github.com/kubernetes/kubernetes/pull/126977 is to make it so that `APILifecycleRemoved` & `APILifecycleDeprecated` is non-blocking and only warns.  This would mean modifying logic in the two methods APILifecycleRemoved is currently called in:
- staging/src/k8s.io/apiserver/pkg/endpoints/deprecation/deprecation.go
- staging/src/k8s.io/apiserver/pkg/server/deleted_kinds.go

### What did you expect to happen?

I expected to be able to bump DefaultKubeBinaryVersion s/v1.31/v1.32 without any integration test failures.

### How can we reproduce it (as minimally and precisely as possible)?

Modify the code [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/version/base.go#L69) to bump DefaultKubeBinaryVersion and then run the kubernetes integration tests.  For a quick validation, you can run: test/integration/auth/selfsubjectreview_test.go `TestGetsSelfAttributes`

**Comment by aaron-prindle** [comment]
From a quick investigation into making the changes necessary to make the generated APILifecycleRemoved method non blocking (quick WIP Branch here: https://github.com/aaron-prindle/kubernetes/tree/fix-127249)

Most relevant commits from above branch:
- [attempt to remove instances where APILifecycleRemoved is used to remove the API (vs warn)](https://github.com/kubernetes/kubernetes/commit/3abf19662a496f39cbd602aaa5eda79eeb76eb94)
- [comment out tests that broke to see show what all breaks](https://github.com/kubernetes/kubernetes/commit/97a9850fcc4802f47e26f8b645fd982b47dcc1eb)


It seems that the changes from the branch/commits shown above would essentially be undoing the changes from this PR from ~4 yrs ago:
https://github.com/kubernetes/kubernetes/pull/96525

When the short-term solution of making APILifecycleRemoved non-blocking was initially discussed, there was an understanding that we would be undoing a recent change but it seems actually this change was made prior to our initial thoughts.  With this new understanding the currently proposed short-term solution is potentially less desirable.  

This means we should discuss additional ways to on how we can address the current set of APIs that are in APILifecycleRemoved but are not removed as they are blocking the PR to bump `DefaultKubeBinaryVersion` 1.31 -> 1.32.  

Currently the options that have been discussed include:
- increase the version gap that is automatically added for APIs via `// +k8s:prerelease-lifecycle-gen:introduced=<version>`. 
 Currently it is defined-version+6 minor versions.  (related code here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/prerelease-lifecycle-gen/prerelease-lifecycle-generators/status.go#L413)
- make `APILifecycleRemoved` non-blocking and only WARN (see info/branch above for more details and info)
- reach out to the API owners for all ~4-6 API gruops mentioned above and coordinate with them removing their APIs.  NOTE: this is likely not possible to be done in the time frame necessary for the PR to bump `DefaultKubeBinaryVersion` as it is blocking additional PRs/workstream (namely those related to adding/promoting feature gates)

I will reach out and discuss these options with @deads2k, @liggitt and @jpbetz for now and put the conclusion in the thread here.

**Comment by liggitt** [comment]
> It seems that the changes from the branch/commits shown above would essentially be undoing the changes from this PR from ~4 yrs ago:
> https://github.com/kubernetes/kubernetes/pull/96525

That PR added the directive to beta APIs.

The 6-version policy currently applies only to beta APIs. My understanding is that the problems are being encountered in alpha APIs which were not in scope for the current policy. I would start by:
1. adjusting the generated policy *for alpha APIs only* to unblock this bump
2. follow-up to propose applying the expiration policy to alpha APIs
3. place specific deprecated / removal versions on the pre-existing alpha APIs in conjunction with 2 and with feedback from the owners of those APIs

**Comment by liggitt** [comment]
I think limiting the auto-added removed minor to beta APIs in https://github.com/kubernetes/kubernetes/blob/d55b119d34883bbad2a3436dcb6c62339d963031/staging/src/k8s.io/code-generator/cmd/prerelease-lifecycle-gen/prerelease-lifecycle-generators/status.go#L422 would be a good step until auto-removal of alpha APIs is accepted
